src/SharedMemoryParcel.php

```diff
@@ -0,0 +1,481 @@
+            throw new \Error('Invalid permissions');
+        }
+
+        $parcel = new self(0, $mutex, $serializer);
+        $parcel->init($value, $size, $permissions);
+        return $parcel;
+    }
+
+    /**
+     * @param Mutex $mutex
+     * @param int $key
+     * @param Serializer|null $serializer
+     *
+     * @return self
+     *
+     * @throws ParcelException
+     */
+    public static function use(Mutex $mutex, int $key, ?Serializer $serializer = null): self
+    {
+        $parcel = new self($key, $mutex, $serializer);
+        $parcel->open();
+        return $parcel;
+    }
+
+    private static function makeKey(string $id): int
+    {
+        return (int) \abs(\unpack("l", \md5($id, true))[1]);
+    }
+
+    /** @var int The shared memory segment key. */
+    private int $key;
+
+    /** @var Mutex A mutex for synchronizing on the parcel. */
+    private Mutex $mutex;
+
+    /** @var \Shmop An open handle to the shared memory segment. */
+    private ?\Shmop $handle = null;
+
+    private int $initializer = 0;
+
+    private Serializer $serializer;
+
+    /**
+     * @param int $key
+     * @param Serializer|null $serializer
+     */
+    private function __construct(int $key, Mutex $mutex, ?Serializer $serializer = null)
+    {
+        if (!\extension_loaded("shmop")) {
+            throw new \Error(__CLASS__ . " requires the shmop extension");
+        }
+
+        $this->key = $key;
+        $this->mutex = $mutex;
+        $this->serializer = $serializer ?? new NativeSerializer;
+    }
+
+    public function getKey(): int
+    {
+        return $this->key;
+    }
+
+    public function unwrap(): mixed
+    {
+        $lock = $this->mutex->acquire();
+
+        try {
+            return $this->getValue();
+        } finally {
+            $lock->release();
+        }
+    }
+
+    public function synchronized(\Closure $closure): mixed
+    {
+        $lock = $this->mutex->acquire();
+
+        try {
+            $result = $closure($this->getValue());
+
+            if ($result !== null) {
+                $this->wrap($result);
+            }
+        } finally {
+            $lock->release();
+        }
+
+        return $result;
+    }
+
+    /**
+     * Frees the shared object from memory.
+     *
+     * The memory containing the shared value will be invalidated. When all
+     * process disconnect from the object, the shared memory block will be
+     * destroyed by the OS.
+     */
+    public function __destruct()
+    {
+        if ($this->initializer === 0 || $this->initializer !== \getmypid()) {
+            return;
+        }
+
+        if ($this->isFreed()) {
+            return;
+        }
+
+        // Invalidate the memory block by setting its state to FREED.
+        $this->writeSegment(self::generateHeader(self::STATE_FREED, 0, 0));
+
+        // Request the block to be deleted, then close our local handle.
+        $this->deleteSegment();
+        $this->handle = null;
```

Comment:
So the original creator is always the real owner of the segment?

Reply:
AFAIK as long as any process has a handle to the segment the OS will keep it around. So there isn't an 'owner' per-say.

Reply:
Yeah, but the original owner will write the `STATE_FREED` header, no?

Reply:
Yes, you're right. Not sure why there's an explicit call to unset mutex below either…